### ISSUE TITLE
Add procedure `+` overload for List and LinkedList

### ISSUE THREAD

**Issue by souris-dev** [chapel-lang/chapel#15373]
### Summary of Problem
I think it would be nice to have an overload of the `+` operator for lists and linked lists that will return a copy that is the result of the second operand being appended to the first one, without changing the contents of the operands (like the `+` for lists in python). 

Something that would allow this:
``` chapel
use List;

var l1: list(int) = [1, 2, 3];
var l2: list(int) = [4, 5, 6];

var l3 = new list(int);
l3 = l1 + l2;

writeln(l3);
writeln(l1);
writeln(l2);
```

This should output:
```
[1, 2, 3, 4, 5, 6]
[1, 2, 3]
[4, 5, 6]
```

**Comment by rahulghangas** [comment]
By chapel philosophy, the ```+``` operator in this case should be promoted to element-wise addition. So, I think that would be more suited as an operator overload.

Also, an extend procedure already exists which appends a list by the elements of another one..

**Comment by souris-dev** [comment]
The `+` operator for element-wise operation would not make sense if the list is of a type that does not have the addition operation defined for them, for example, in case of record types that cannot be "added" per se.

The extend procedure modifies the object on which it is called. What I am proposing, however, is something that would let us have the result of concatenation of the two operands without changing the operands themselves. In the above example, `l1` and `l2` are unchanged. However, doing `l1.extend(l2)` would change l1 itself.

What I am proposing can, of course, currently be currently done like this:
``` chapel
...
var l3 = new list(int);
l3.extend(l1);
l3.extend(l2);
```
I believe that using the `+` operator to do the same would be more intuitive.

**Comment by ty1027** [comment]
Doesn't the `+` operator become possibly confusing when used for lists and arrays at the same time? The following example (in Python) may be a bit artificial, but I am afraid similar confusion might happen even for Chapel (if the meaning of `+`is extended more and generics are involved).

```python
import numpy as np

a = [1, 2]    # python list
b = [3, 4]

print( a + b )      # [1, 2, 3, 4]
print( a + b + a )  # [1, 2, 3, 4, 1, 2]

x = np.array([100, 200])  # numpy ndarray

print( a + x )  # [101, 202]
print( x + a )  # [101, 202]

# print( a + b + x )
# print( (a + b) + x )
# print( x + (a + b) )
# Error: operands could not be broadcast together with shapes (4,) (2,) 

print( x + a + b )   # [104 206]
print( a + (b + x) )  # [104 206]
```

For comparison, some other (statically typed) languages use different symbols for concatenation (like `~` in D and `&` in Nim), which may be nice to avoid possible confusion.

https://dlang.org/spec/arrays.html#array-concatenation 
[(TIO)](https://tio.run/##bY/BagMhEIbv@xT/LbvUJnSz9CJ5ksWDqSYdUCeoNITQvPp2TKCk0AEdxW@@Gd2yUDxxrijVrWUR6677YnKIllI/dNcOEpTqbGCxw/ymMBqFfTtvFSYjBQ05Z6o@pF6om7wOGvfYbFrJqLZqMv9xN8mNfeKUZNM99XX6fnHr4NOxfkrnUT/UNgT@sNUj@sj5ggNnrNwKfaFIwWZUhi2Fjin6VMEHOG4fGx5Cce9gZX/Bfjb673zud65JvRv0PvgmeT1T8bDOUSUW0fey/AA)

https://nim-lang.org/docs/sequtils.html   (where "seq" is similar to Python list)
[(TIO)](https://tio.run/##y8vM/f@/LLFIIVHBVsEh2lDHKJYLxE0Cc411TGK5UpMz8jWA8mpAQU0FBWWIMh2QnIJGokJiXgpQIju1oEQhM68kMbkEqIYw4ErUS0xJ0YAYqQAyNVEhs1ghNz8lMy0zNQVolG5BTmJyKtx2qKnItv//DwA)

Apart from the choice of symbol, I am also wondering whether a function approach (e.g. `concat(a, b, c)` where `concat()` takes variadic arguments) might be more efficient than a repeated use of a binary operator (e.g. `a ~ b ~ c`) if allocation of temporary variables could be done only once.

https://nim-lang.org/docs/sequtils.html#concat%2Cvarargs%5Bseq%5BT%5D%5BT%5D%5D
[(TIO)](https://tio.run/##y8vM/f8/M7cgv6hEoTi1sLQkM6eYi6sssUghUcFWwSHaUEfBKBbMTwLzjXUUTGK5uFKTM/I1FJLz85ITSzQUEnUUknSAGjQVNP//BwA)

(this is a bit different but for comparison...)
https://docs.julialang.org/en/v1/base/arrays/#Base.vcat
[(TIO)](https://tio.run/##bY47DsIwEER7n2IkmqwyFHYCouEkkQsHuTAyFgLzub1xDAVCTLNTPL2d4y0Gp5@lOOwxacJYNS91IEar1PkSUo6pg0OPGYKWFaaR2Fp0PvqTT3n9CFcvv3RfrzR6Q@y@bfeDy5ViU8pHqWk4sH79g7Gp5I3BEMs@og0u5QU)

**Comment by Rapiz1** [comment]
I agree with @ty1027 .  This feature does not bring significant convenience to users but brings confusion and potential performance issues.

**Comment by ty1027** [comment]
For more clarification, I think it would be nice to have some syntax like "a [?] b" (where [?] is some operator for list concatenation) in addition to a function like concat( a, b ). But I am not sure if the symbol [?] = "+" is really a good choice here... (though Python is using it). This is partly because some other languages seem to avoid it, e.g.

https://alvinalexander.com/scala/how-to-merge-concatenate-lists-in-scala-cookbook/
https://stackoverflow.com/a/32135472

Here, I don't think the operator ":::" in the above page is very appealing XD, but the StackOverflow answer above (which mentions type safety) seems to be also useful here.

**Comment by souris-dev** [comment]
The operator certainly could be different from `+`, but I think having some operator or a function like `concat()` (as suggested by @ty1027) would be quite convenient. That would eliminate any confusion caused by `+`.

For now, we could go with a `concat()` defined in the `list` and `LinkedList` modules that takes variadic arguments.

**Comment by ArtemSarmini** [comment]
`LinkedList` already has `concat` method, which is analog of `extend` from `list`. Shouldn't these have the same name `extend`?

**Comment by e-kayrakli** [comment]
@ArtemSarmini is right. We already have `list.extend` for extending a list by another list.

@souris-dev -- Do you think `extend` is OK? Do you think `concat` would be a better name? Note that our standard modules (e.g. List) gets more use and attention from users and developers compared to package modules (e.g. LinkedList). So, design-wise, I'd say the interface in the standard modules (`extend`) represents what we want more closely than that in the package modules (`concat`).

@ty1027 -- How strongly do you feel about having an operator for this vs using a method? Keep in mind that unless we use an [existing overloadable operator](https://chapel-lang.org/docs/master/language/spec/procedures.html#function-and-operator-overloading), we'll need to figure out whether adding a new operator to the language make sense in the big picture.

My personal opinions; I completely agree that `+` is not the right operator for this, because of reasons already mentioned. I cannot think of any of the existing operators to be sufficient in expressing what `list1.extend(list2)` does. Nor I personally feel the need for it.

**Comment by ty1027** [comment]
Hi @e-kayrakli , thank you for your message, and I am totally fine with a method approach (like .append() / .extend()), and using a function like concat() (if necessary) to create a new list from other ones (w/o modifying the latter). I guess the latter could also be written by users rather directly (if plain implementation is fine).

RE new set of operators, I am afraid it will take significant development time, while 
providing not so much actual benefit to user coding... Non-standard operators/symbols may make the program more concise and possibly allow expressions closer to
math formulas, but IMO it also tends to make the program harder
to understand when the reader is not familiar to their meaning (i.e., code becomes
more "cryptic"). Introduction of some very limited new symbols might be useful
(not for user but for core syntax), but I am afraid allowing the use of "general"
symbols (like UTF) for user-defined operators is not so useful (just my opinion though).

**Comment by souris-dev** [comment]
@e-kayrakli I agree with @ArtemSarmini. As you said, it would make more sense to have both of them named `extend()`.

Also, could we have another method (appropriately named) to not modify the list in-place but rather just return a resultant extended list? An operator may not be necessary for this, but I think such a method could be of convenience.
